IPlug/CLAP/IPlugCLAP.cpp

```diff
@@ -0,0 +1,1015 @@
+{
+  const IParam* pParam = GetParam(idx);
+  const bool isDoubleType = pParam->Type() == IParam::kTypeDouble;
+  const double value = isDoubleType ? normalizedValue : pParam->FromNormalized(normalizedValue);
+  
+  mParamValuesToHost.PushFromArgs(ParamToHost::Type::Value, idx, value);
+}
+
+void IPlugCLAP::EndInformHostOfParamChange(int idx)
+{
+  mParamValuesToHost.PushFromArgs(ParamToHost::Type::End, idx, 0.0);
+}
+
+bool IPlugCLAP::EditorResize(int viewWidth, int viewHeight)
+{
+  if (HasUI())
+  {
+    if (viewWidth != GetEditorWidth() || viewHeight != GetEditorHeight())
+    {
+      GetClapHost().guiRequestResize(viewWidth, viewHeight);
+    }
+    
+    SetEditorSize(viewWidth, viewHeight);
+  }
+  
+  return true;
+}
+
+// IPlugProcessor
+void IPlugCLAP::SetTailSize(int samples)
+{
+  IPlugProcessor::SetTailSize(samples);
+  
+  if (GetClapHost().canUseTail())
+    mTailUpdate = true;
+}
+
+void IPlugCLAP::SetLatency(int samples)
+{
+  IPlugProcessor::SetLatency(samples);
+  
+  if (GetClapHost().canUseLatency())
+  {
+    // If active request restart on the main thread (or update the host)
+    if (isActive())
+    {
+      mLatencyUpdate = true;
+      runOnMainThread([&](){ if (!isBeingDestroyed()) GetClapHost().requestRestart(); });
+    }
+    else
+    {
+      runOnMainThread([&](){ if (!isBeingDestroyed()) GetClapHost().latencyChanged(); });
+    }
+  }
+}
+
+bool IPlugCLAP::SendMidiMsg(const IMidiMsg& msg)
+{
+  mMidiToHost.Add(msg);
+  return true;
+}
+
+bool IPlugCLAP::SendSysEx(const ISysEx& msg)
+{
+  // TODO - don't copy the data
+  SysExData data(msg.mOffset, msg.mSize, msg.mData);
+  mSysExToHost.Add(data);
+  return true;
+}
+
+// clap_plugin
+bool IPlugCLAP::init() noexcept
+{
+  SetDefaultConfig();
+  
+  return true;
+}
+
+bool IPlugCLAP::activate(double sampleRate, uint32_t minFrameCount, uint32_t maxFrameCount) noexcept
+{
+  SetBlockSize(maxFrameCount);
+  SetSampleRate(sampleRate);
+  OnActivate(true);
+  OnParamReset(kReset);
+  OnReset();
+
+  mHostHasTail = GetClapHost().canUseTail();
+  mTailCount = 0;
+
+  return true;
+}
+
+void IPlugCLAP::deactivate() noexcept
+{
+  OnActivate(false);
+  
+  if (mLatencyUpdate)
+  {
+    GetClapHost().latencyChanged();
+    mLatencyUpdate = false;
+  }
+
+  // TODO - should we clear mTailUpdate here or elsewhere?
+}
+
+template <typename T>
+const T* ClapEventCast(const clap_event_header_t* event)
+{
+  return reinterpret_cast<const T*>(event);
+}
+
+template <typename T>
+bool InputIsSilent(const T* data, int nFrames)
+{
+  // For local tail processing find non-zero inputs
+  auto isZero = [](T x) { return x == T(0); };
+
+  return std::find_if_not(data, data + nFrames, isZero) == (data + nFrames);
+}
+
+clap_process_status IPlugCLAP::process(const clap_process* process) noexcept
+{
+  IMidiMsg msg;
+  SysExData sysEx;
+  
+  // Transport Info
+  if (process->transport)
+  {
+    ITimeInfo timeInfo;
+
+    auto &transport = process->transport;
```

Comment:
not immediately clear why auto& here?

Reply:
I can't see a reason - it's probably easier for the compiler to optimise a stack copy of the pointer, so this looks either like a mistake, or just that I was thinking about it weirdly when I did it.